sycl/doc/extensions/proposed/sycl_ext_oneapi_reduction_properties.asciidoc

```diff
@@ -0,0 +1,228 @@
+                          PropertyList properties);
+
+template <typename BufferT, typename BinaryOperation, typename PropertyList>
+__unspecified__
+reduction(BufferT vars, handler& cgh, const BufferT::value_type& identity,
+          BinaryOperation combiner, PropertyList properties);
+
+template <typename T, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(T* var, const T& identity, BinaryOperation combiner,
+                          PropertyList properties);
+
+template <typename T, typename Extent, typename BinaryOperation, typename PropertyList>
+__unspecified__ reduction(span<T, Extent> vars, const T& identity,
+                          BinaryOperation combiner,
+                          PropertyList properties);
+
+}
+----
+
+=== Reduction properties
+
+New `reduction` properties are introduced to allow developers to constrain
+reduction algorithm selection based on desired behavior(s). Compile-time
+properties corresponding to existing runtime properties are also introduced to
+ensure that all information can be passed via a single property list.
+
+If a reduction kernel is submitted to a device that cannot satisfy the
+request for specific reduction behavior(s), the implementation must throw an
+`exception` with the `errc::feature_not_supported` error code.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+struct deterministic_key {
+  using value_t = property_value<deterministic_key>;
+};
+inline constexpr deterministic_key::value_t deterministic;
+
+struct initialize_to_identity_key {
+  using value_t = property_value<initialize_to_identity_key>;
+};
+inline constexpr initialize_to_identity_key::value_t initialize_to_identity;
+
+}
+----
+
+|===
+|Property|Description
+
+|`deterministic`
+|Adds the requirement that any two calls to the reduction on the same device
+which use the same launch configuration (i.e., `range` or `nd_range`) and in
+which the same kernel instances contribute the same values to the reduction
+will produce results with the same value representation.
+
+[_Note:_ Kernel instances may contribute different values to a reduction
+because of other potential sources of non-determinism, such as calls to group
+algorithms, use of atomic operations, etc. _{endnote}_]
+
+|`initialize_to_identity`
+|Adds the same requirement as
+`sycl::property::reduction::initialize_to_identity`.
+
+|===
+
+
+=== Usage example
+
+[source,c++]
+----
+float sum(sycl::queue q, float* input, size_t N) {
+
+  float result = 0;
+  {
+    sycl::buffer<float> buf{&result, 1};
+
+    q.submit([&](sycl::handler& h) {
+      auto reduction = sycl::reduction(buf, h, sycl::plus<>(), sycl::properties(sycl::deterministic));
+      h.parallel_for(N, reduction, [=](size_t i, auto& reducer) {
+        reducer += input[i];
+      });
+    }
+  }
+  return result;
+  
+}
+
+...
+
+float x = sum(q, array, 1024);
+float y = sum(q, array, 1024);
+assert(sycl::bit_cast<unsigned int>(x) == sycl::bit_cast<unsigned int>(y));
```

Comment:
I don't understand the problem that needs to be solved here.  I noticed that the description says that the property guarantees the same "value **representation**" not the same "value", and I see that you are using `bit_cast` here.  Therefore, I assume you are distinguishing between "value" and its "representation".

How could this reduction produce two identical values, but with different representations?  Are you worried about positive vs. negative zero, for example?  Or, different flavors of NaN?

Reply:
I might have overcomplicated this.  I've swapped "value representation" for "value" and tweaked some other language in de8c5a79959056d66b3c1ac402ca3997ae18886d.  The results of the reduction should be identical.

I'm using `bit_cast` here for two reasons.  First, I wanted to highlight that `deterministic` guarantees bitwise reproducible results for floating-point types: this is important for multiple domains (e.g., climate simulation, training generative AI), and it's probably the main use-case for this property.  Second, I wanted to sidestep the issue that `assert(x == y)` would fail if the result was NaN.